lib/internal/mime.js

```diff
@@ -16,62 +16,140 @@ const {
+    }
+  }
+
+  marker.position = index;
+
+  // #6 If the extract-value flag is set, then return value
+  return value;
+}
+
 function parseTypeAndSubtype(str) {
-  // Skip only HTTP whitespace from start
-  let position = SafeStringPrototypeSearch(str, END_BEGINNING_WHITESPACE);
-  // read until '/'
-  const typeEnd = StringPrototypeIndexOf(str, SOLIDUS, position);
-  const trimmedType = typeEnd === -1 ?
-    StringPrototypeSlice(str, position) :
+  const strLength = str.length;
+  // #2 Let position be a position variable for input, initially pointg at
+  // the start of the input
+  let position = 0;
+
+  // #3 Let type be the result of collecting a sequence of code points that are
+  // not U+002F (/ [SOLIDUS]) from input, given position.
+  const typeEnd = StringPrototypeIndexOf(str, SOLIDUS);
+  const preType = typeEnd === -1 ?
+    str :
     StringPrototypeSlice(str, position, typeEnd);
-  const invalidTypeIndex = SafeStringPrototypeSearch(trimmedType,
-                                                     NOT_HTTP_TOKEN_CODE_POINT);
-  if (trimmedType === '' || invalidTypeIndex !== -1 || typeEnd === -1) {
-    throw new ERR_INVALID_MIME_SYNTAX('type', str, invalidTypeIndex);
+
+  // #4 If type is the empty string or does not solely
+  // contain HTTP token code points, then return failure.
+  // #5 If position is past the end of input, then return failure.
+  if (typeEnd === -1 || preType.length === 0 || typeEnd > strLength || SafeStringPrototypeSearch(
+    preType, NOT_HTTP_TOKEN_CODE_POINT) !== -1) {
+    throw new ERR_INVALID_MIME_SYNTAX('type', str, typeEnd);
   }
-  // skip type and '/'
+
+  // #6 Advance position by 1. (This skips past U+002F (/).)
   position = typeEnd + 1;
-  const type = toASCIILower(trimmedType);
-  // read until ';'
-  const subtypeEnd = StringPrototypeIndexOf(str, SEMICOLON, position);
-  const rawSubtype = subtypeEnd === -1 ?
-    StringPrototypeSlice(str, position) :
-    StringPrototypeSlice(str, position, subtypeEnd);
+
+  // #7 Let subtype be the result of collecting a sequence of code points
+  // that are not U+003B (;) from input, given position.
+  let subtypeEnd = StringPrototypeIndexOf(str, SEMICOLON, position);
+  subtypeEnd = subtypeEnd === -1 ? strLength : subtypeEnd;
+  const rawSubtype = StringPrototypeSlice(str, position, subtypeEnd);
   position += rawSubtype.length;
-  if (subtypeEnd !== -1) {
-    // skip ';'
-    position += 1;
-  }
-  const trimmedSubtype = StringPrototypeSlice(
-    rawSubtype,
-    0,
-    SafeStringPrototypeSearch(rawSubtype, START_ENDING_WHITESPACE));
-  const invalidSubtypeIndex = SafeStringPrototypeSearch(trimmedSubtype,
-                                                        NOT_HTTP_TOKEN_CODE_POINT);
-  if (trimmedSubtype === '' || invalidSubtypeIndex !== -1) {
-    throw new ERR_INVALID_MIME_SYNTAX('subtype', str, trimmedSubtype);
+
+  // #8 Remove any trailing HTTP whitespace from subtype.
+  const preSubType = StringPrototypeTrimEnd(rawSubtype);
```

Comment:
can we avoid the trim call?
can we avoid a new variable?

Reply:
Hmm, not without iterating through the string to look for trailing whitespace, otherwise it is also counted as char and can lead to false positives.

On the other hand, the variable allocation can be removed, and mounted within the `if` statement 🤔

Reply:
The thing is, that you actually need to check if rawSubtype.charCodeAt(rawSubtype.length - 1) is a whitespace. If so, you should trim. Maybe trimEnd is optimized enough. But allocating a new variable is a pot. perf. bottleneck.